akka-persistence-typed/src/test/java/jdocs/akka/persistence/typed/auction/AuctionEntity.java

```diff
@@ -0,0 +1,241 @@
+   */
+  private Effect<AuctionEvent, AuctionState> handleAutomaticOutbid(
+      PlaceBid bid, Auction auction, Instant now, Optional<Bid> currentBid, int currentBidPrice, int currentBidMaximum) {
+    // Adjust the bid so that the increment for the current maximum makes the current maximum a valid bid
+    int adjustedBidPrice = Math.min(bid.getBidPrice(), currentBidMaximum - auction.getIncrement());
+    int newBidPrice = adjustedBidPrice + auction.getIncrement();
+
+    return Effect().persist(Arrays.asList(
+        new BidPlaced(entityUUID,
+            new Bid(bid.getBidder(), now, adjustedBidPrice, bid.getBidPrice())
+        ),
+        new BidPlaced(entityUUID,
+            new Bid(currentBid.get().getBidder(), now, newBidPrice, currentBidMaximum)
+        )
+      ))
+      .thenReply(bid, newState -> new PlaceBidResult(PlaceBidStatus.ACCEPTED_OUTBID, newBidPrice, currentBid.get().getBidder()));
+  }
+
+  /**
+   * Handle the situation where a bid will be accepted as the new winning bidder.
+   */
+  private Effect<AuctionEvent, AuctionState> handleNewWinningBidder(PlaceBid bid,
+                                                       Auction auction, Instant now, int currentBidMaximum) {
+    int nextIncrement = Math.min(currentBidMaximum + auction.getIncrement(), bid.getBidPrice());
+    int newBidPrice;
+    if (nextIncrement < auction.getReservePrice()) {
+      newBidPrice = Math.min(auction.getReservePrice(), bid.getBidPrice());
+    } else {
+      newBidPrice = nextIncrement;
+    }
+    return Effect().persist(new BidPlaced(
+        entityUUID,
+        new Bid(bid.getBidder(), now, newBidPrice, bid.getBidPrice())
+      ))
+      .thenReply(bid, newState -> {
+        PlaceBidStatus status;
+        if (newBidPrice < auction.getReservePrice()) {
+          status = PlaceBidStatus.ACCEPTED_BELOW_RESERVE;
+        } else {
+          status = PlaceBidStatus.ACCEPTED;
+        }
+        return new PlaceBidResult(status, newBidPrice, bid.getBidder());
+      });
+  }
+
+  @Override
+  public AuctionState emptyState() {
+    return AuctionState.notStarted();
+  }
+
+  @Override
+  public CommandHandler<AuctionCommand, AuctionEvent, AuctionState> commandHandler() {
+    return notStartedHandler
+        .orElse(underAuctionHandler)
+        .orElse(completedHandler)
+        .orElse(cancelledHandler)
+        .orElse(getAuctionHandler)
+        .orElse(cancelledHandler)
+        .build();
+  }
+
+  @Override
+  public EventHandler<AuctionState, AuctionEvent> eventHandler() {
+    return eventHandlerBuilder()
+        .matchEvent(AuctionStarted.class, (state, evt) -> AuctionState.start(evt.getAuction()))
+        .matchEvent(BidPlaced.class, (state, evt) -> state.bid(evt.getBid()))
+        .matchEvent(BiddingFinished.class, (state, evt) -> state.withStatus(AuctionStatus.COMPLETE))
+        .matchEvent(AuctionCancelled.class, (state, evt) -> state.withStatus(AuctionStatus.CANCELLED))
+        .build();
```

Comment:
On another experiment, I realised that this style is less safe. As defined here, any event can be applied at any moment. We are assuming that the command handlers are always correct and won't emit 'bad' events. 

This is in principle correct. If the event is persisted, it should be legal to apply it. However, as a safety measure, we first apply the events before persisting it to guarantee that we have an event handler for it. With the above, we always have an event handler, disregard in which state is the auction.

Reply:
Yes, in principle you should have validated the command before persisting it so applying should always succeed. I agree that it can be a good practice to have an additional check in the event handler.

There is already an overloaded `matchState` with a `stateClass` parameter for that purpose. Perhaps we need the predicate also, although not as important as for command handlers where it serves the purpose of fall through to next match case. Exception can be thrown from inside the lambda also.

Reply:
I think we should align it with the style in `CommandHandlerBuilder`. Starting from a `State` and then defining "matches" for the events. 

We should do that for a few reasons: 

1. API symetry. If I learn how to build one, I know how to build the other. 
1. It reads well: "given that state, when that event comes in, do that"
1. We need the means to define handlers when the state is null (initial state). That's not possible with the variant where we pass a `stateClass`.
1. In principle, I find it strange to have a predicate, other then the special case of null check, but may make sense in Java. Java devs may prefer to use a status enum instead of an ADT.